File: PHP/dynamic_programming/rod_cutting.php
<?php

/**
 * Given a rod of length N inches and a table of prices P for i=1,2,...n,
 * determine the maximum revenue R obtainable by cutting up the rod and selling
 * the pieces. 
 * Note that if the price P for a rod length n is large enough, an optimal solution
 * may require no cutting at all.
 */

$price = [0,1,5,8,9,10,17,17,20,24,30];

/**
 * For each rod length, computations required are:
 * cutrod(4) -> cutrod(3), cutrod(2), cutrod(1), cutrod(0)
 * cutrod(3) -> cutrod(2), cutrod(1), cutrod(0)
 * cutrod(2) -> cutrod(1), cutrod(0)
 * cutrod(1) -> cutrod(0)
 *
 * Time complexity: O(2^n)
 * This algorithm explicitly considers the 2^(n-1) ways of cutting up the rod.
 * Computing each leaf explicitly.
 */
function cutRod($price, $rodLength) {
    if ($rodLength == 0) {
        return 0;
    }
    $revenue = 0;
    for ($i=1; $i <= $rodLength ; $i++) { 
        $revenue = max($revenue, $price[$i] + cutRod($price, $rodLength-$i));
    }
    return $revenue;
}

/**
 * Top down approach with memoization.
 * At each computation, we check if the computation has already been computed.
 * If yes, we just use the computed value. Else, we compute it.
 *
 * Time complexity: O(n^2)
 * 
 */
function cutRodMemoized($price, $rodLength) {
    $revenueTable = array_fill(0, $rodLength+1, -1);
    return _cutRodMemoized($price, $rodLength, $revenueTable);
}

function _cutRodMemoized($price, $rodLength, $revenueTable) {
    if ($rodLength == 0) {
        return 0;
    } elseif ($revenueTable[$rodLength] >= 0) {
        return $revenueTable[$rodLength];
    }
    $revenue = -1;
    for ($i=1; $i <= $rodLength; $i++) { 
        $revenue = max($revenue, $price[$i] + _cutRodMemoized($price, $rodLength-$i, $revenueTable));
    }
    $revenueTable[$rodLength] = $revenue;
    return $revenue;
}

/**
 * Bottom up approach.
 *
 * Time complexity: O(n^2)
 */
function cutRodV3($price, $rodLength) {
    $revenueTable = array_fill(0, $rodLength+1, -1);
    $optimalSize = array_fill(0, $rodLength+1, -1);
    $revenueTable[0] = 0;
    for ($i=1; $i <= $rodLength ; $i++) {
        for ($j=1; $j <= $i ; $j++) {
            if ($revenue < $price[$j] + $revenueTable[$i-$j]) {
                $revenue = $price[$j] + $revenueTable[$i-$j];
                $optimalSize[$i] = $j;
            }
        }
        $revenueTable[$i] = $revenue;
    }
    echo "rod length: " . $rodLength . 
        " revenue: " . $revenueTable[$rodLength] .
        " solution : " . implode(" ", getCutRodSolution($rodLength, $optimalSize)) . PHP_EOL;
    return $revenueTable[$rodLength];
}

function getCutRodSolution($rodLength, $optimal) {
    $result = [];
    $start = $rodLength;
    while ($start > 0) {
        $result[] = $optimal[$start];
        $start -= $optimal[$start];
    }
    return $result;
}

cutRodV3($price, 1);
cutRodV3($price, 2);
cutRodV3($price, 3);
cutRodV3($price, 4);
cutRodV3($price, 5);
cutRodV3($price, 6);
cutRodV3($price, 7);
cutRodV3($price, 8);
cutRodV3($price, 9);
cutRodV3($price, 10);

File: Python/dynamic_programming/rod_cutting.py
# Serling Enterprises buys long steel rods and cuts them into shorter rods,
# which it then sells. Each cut is free. The management of Serling Enterprises
# wants to know the best way to cut up the rods.

# Given a rod of length n inches and a table of prices for rod length
# i = 1, 2,...,n. Determine the maximum revenue obtainable by cutting up the
# rod and selling the pieces. Note that if the price Pn for a rod of length n
# is large enough, an optimal solution may require no cutting at all.

# length i | 1 2 3 4 5  6  7  8  9  10
# price Pi | 1 5 8 9 10 17 17 20 24 30

# Consider the case when n = 4. Cutting a 4 inch rod into 2 inch pieces
# produces revenue P2 + P2 = 5 + 5 = 10, which is optimal.


def cut_rod_recursive(price_table, n):
    """
    Recursive solution to cut rod problem. Takesn in a price_table and a number
    n for which to calculate the optimal revenue.

    """
    if n == 0:
        return 0
    result = -1
    # go through all possible ways of cutting the rod of current length and get
    # the max revenue of these values
    for i in xrange(1, n+1):
        result = max(result, price_table[1][i] + cut_rod_recursive(price_table,
                                                                   n-i))
    return result

# Dynamic programming solution by top down apporach


def cut_rod_memoized(price_table, n):
    # initialize DP table
    dp_table = [-1 for i in xrange(n+1)]
    return cut_rod_memoized_aux(price_table, n, dp_table)


def cut_rod_memoized_aux(price_table, n, dp_table):
    if dp_table[n] >= 0:
        return dp_table[n]
    if n == 0:
        result = 0
    else:
        result = -1
        for i in xrange(1, n+1):
            result = max(result, price_table[1][i] +
                         cut_rod_memoized_aux(price_table, n-i, dp_table))
    dp_table[n] = result
    return result

# Dynamic programming solution by bottom-up method


def cut_rod_bottom_up(price_table, n):
    """
    In DP bottom-up approach, we create a solution table to store the size of
    the first cut in each of the optimal sub solutions and return the actual
    solution of cutting a rod of n size.

    """
    dp_table = [-1 for i in xrange(n+1)]
    solutions = [-1 for i in xrange(n+1)]
    # initialize first term
    dp_table[0] = 0
    solutions[0] = 0
    for i in xrange(1, n+1):
        temp = -1
        for j in xrange(1, i+1):
            if temp < price_table[1][j] + dp_table[i-j]:
                temp = max(temp, price_table[1][j] + dp_table[i-j])
                solutions[i] = j
        dp_table[i] = temp
    return dp_table[n], (solutions[n], n - solutions[n])


if __name__ == "__main__":
    table = [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
             [0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30]]

    print cut_rod_bottom_up(table, 0)
    print cut_rod_bottom_up(table, 1)
    print cut_rod_bottom_up(table, 2)
    print cut_rod_bottom_up(table, 3)
    print cut_rod_bottom_up(table, 4)
    print cut_rod_bottom_up(table, 5)
    print cut_rod_bottom_up(table, 6)
    print cut_rod_bottom_up(table, 7)
    print cut_rod_bottom_up(table, 8)
    print cut_rod_bottom_up(table, 9)
    print cut_rod_bottom_up(table, 10)
